### PR TITLE
Add proxy_build_info metric

### PR DESCRIPTION
**Pull request by kleimkuhler**
## Why

Closes linkerd/linkerd2#3424

## What

A new `proxy_build_info` metric is introduced that contains the following
fields:
- `git_branch`: The branch that the proxy was built on
- `git_version`: The version of the proxy. This is the output of `git describe`
  and finds the most recent tag that is reachable from the commit.
- `rust_version`: The rust version used to build the proxy

This metric is a gauge that always has a value of `1`. The reasoning is
described [here](https://www.robustperception.io/exposing-the-software-version-to-prometheus).

The fields reflect those of the `prometheus_build_info` discussed in the link
above.

```bash
❯ linkerd metrics -n linkerd pods/linkerd-tap-5f6565cfc5-56q9r
..
# HELP proxy_build_info Proxy build info
# TYPE proxy_build_info gauge
proxy_build_info{git_branch="kleimkuhler/build-info",git_version="release/v2.104.1-11-gea34a589",rust_version="rustc 1.44.1 (c7087fe00 2020-06-17)",} 1
```

## How

The `.git/` directory is now copied into the docker container when building so
that the `build.rs` file being introduced can reference the git history to find
the most recent tag and populate the `GIT_BRANCH` and `GIT_VERSION` environment
variables at compile time.

### Labels

I considered a `git_release` and `git_revision` tag instead of the single
`git_version`.

`git_release` would be the most recent tag reachable from the commit, and
`git_revision` would be the full commit SHA.

I chose the output of `git describe` instead because it includes both the tag
and short SHA as well as how many additional commits have occurred on top of the
tag.

I think this makes it a little easier as a user to see that the proxy build
version is the `kleimkuhler/build-info` branch, which is `1` commit from
`release/v2.104.1` with the SHA starting with `1ee7452f`:

```bash
❯ git describe
release/v2.104.1-1-g1ee7452f
```

Signed-off-by: Kevin Leimkuhler <kevin@kleimkuhler.com>

### CODE REVIEW

File: linkerd/app/core/build.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+use std::process::Command;
+use std::string::String;
+
+fn main() {
+    let git_branch = match Command::new("git")
+        .args(&["rev-parse", "--abbrev-ref", "HEAD"])
+        .output()
+    {
+        Ok(output) => String::from_utf8(output.stdout).unwrap(),
+        Err(_) => "unavailable".to_string(),
```
[the remark this answers is not in the document]
Comment (hawkw) [reply linkerd/app/core/build.rs:10]: +1, there should probably be a `println!` here.

Note that build scripts can use [`println!("cargo:warning=MESSAGE");`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargo-warning) to make cargo emit a build warning.

File: linkerd/app/core/build.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    };
+    let git_version = match Command::new("git")
+        .args(&["describe", "--always", "HEAD"])
+        .output()
+    {
+        Ok(output) => String::from_utf8(output.stdout).unwrap(),
+        Err(_) => "unavailable".to_string(),
```
[the remark this answers is not in the document]
Comment (hawkw) [reply linkerd/app/core/build.rs:17]: probably worth printing a warning here as well.

File: linkerd/app/core/src/telemetry/build_info.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    // build information in `labels` remains accurate
+    value: Arc<Gauge>,
+    labels: Arc<BuildInfoLabels>,
+}
+
+#[derive(Clone, Debug, Default)]
+struct BuildInfoLabels {
+    git_branch: String,
+    git_version: String,
+    rust_version: String,
```
Comment (hawkw) [inline linkerd/app/core/src/telemetry/build_info.rs:29]: could be worth adding a label for whether the proxy was built in debug mode or release mode (which I believe is [available in an env var called `PROFILE`](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts)).

File: linkerd/app/core/build.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zaharidichev) [reply linkerd/app/core/build.rs:10]: > Note that build scripts can use println!("cargo:warning=MESSAGE"); to make cargo emit a build warning.

Good to know!

[the remark this answers is not in the document]
Comment (kleimkuhler) [reply linkerd/app/core/build.rs:10]: Emitting a build warning sounds good

File: linkerd/app/core/build.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    };
+    let rust_version = match Command::new("rustc").arg("--version").output() {
+        Ok(output) => String::from_utf8(output.stdout).unwrap(),
+        Err(_) => "unavailable".to_string(),
```
[the remark this answers is not in the document]
Comment (kleimkuhler) [reply linkerd/app/core/build.rs]: I now use the error as the message in the build warning.

**Comment by kleimkuhler** [comment]
I've added an additional `profile` label so that the full metric looks like: `proxy_build_info{control_plane_component="controller",control_plane_ns="linkerd",deployment="linkerd-controller",git_branch="kleimkuhler/build-info",git_version="release/v2.104.1-2-gf394bf05",instance="10.42.0.19:4191",job="linkerd-proxy",namespace="linkerd",pod="linkerd-controller-66b798d4d7-9d559",pod_template_hash="66b798d4d7",profile="release",rust_version="rustc 1.44.1 (c7087fe00 2020-06-17)",workload_ns="linkerd"}`

**Pull request merged**